### DISCUSSION TITLE
How to deselect all rows in ui.table

### DISCUSSION THREAD

**Discussion by gbrandt** [zauberzeug/nicegui#150] [Q&A] [answered]
Apologies, I am new at all this...

I cannot figure out how to deselect all the rows in a ui.table.  I see a deselect_rows() call in AgGrid, but have no idea how to call it.  I tried call `ui.run_javascript( "cached_grid_def['gxxx'].api.deselectAll()`  where xxx was my `table.view.id` and that did not work, although it worked in the javascript console in my browser.

Any help would be appreciated.

**Comment by rodja** [comment]
Have you awaited the `ui.run_javascript`-call? It also must be placed in an async function. 

And it can only run *after* the page has been loaded. If you run the javascript call on button click or similar, this is obviously the case. If you need to do this on page creation, try the `yield` mechanism described at https://nicegui.io/reference#yielding_for_page-ready.

**Comment by falkoschindler** [comment]
You can use JustPy's [AgGrid API](https://justpy.io/grids_tutorial/grid_api/) to send commands via `run_api`. Alternatively you can send a plain JavaScript command. Here is an example demonstrating both solutions:

```python
#!/usr/bin/env python3
from nicegui import ui

table = ui.table({
    'columnDefs': [{'field': 'name'}],
    'rowData': [
        {'name': 'Alice'},
        {'name': 'Bob'},
        {'name': 'Carol'},
    ],
    'rowSelection': 'multiple',
    'rowMultiSelectWithClick': True,
}).classes('max-h-40')

async def deselect1():
    await table.view.run_api('deselectAll()', table.page)

async def deselect2():
    await ui.run_javascript(f'cached_grid_def["g{table.view.id}"].api.deselectAll()')

ui.button('Deselect 1', on_click=deselect1)
ui.button('Deselect 2', on_click=deselect2)

ui.run()
```

**Reply by canDry** [reply]
AttributeError: 'Table' object has no attribute 'view'

**Comment by falkoschindler** [comment, marked as answer]
Since NiceGUI 1.0 you can deselect table rows as follows:

```python
ui.button('Deselect', on_click=lambda: table.call_api_method('deselectAll'))
```

**Comment by vlogoth** [comment]
Hello, and apologies for raising this again, but seems entirely relevant to my question.
I was trying to utilize the "clearSelection" method from Quasar to deselect my table.
```py
columns = [
        {'name': 'name', 'label': 'Name', 'field': 'name', 'required': True, 'align': 'left'},
        {'name': 'age', 'label': 'Age', 'field': 'age', 'sortable': True},
    ]
rows = [
        {'name': 'Alice', 'age': 18},
        {'name': 'Bob', 'age': 21},
        {'name': 'Carol'},
    ]
test = ui.table(columns=columns, rows=rows, row_key='name', selection='multiple')
ui.button('lala_1', on_click=lambda: test.__setattr__('selected', []))
ui.button('lala_2', on_click=lambda: test.run_method('clearSelection'))
ui.button('lala_3', on_click=lambda: test.call_api_method('deselectAll'))
```
I was expecting the second option to work as other methods I've used in other components. It doesn't though.
The third one suggested here does not work either, and I get unresolved reference.
The first works just fine but was not as intuitive to reach this working example.
Any takes on this?

**Reply by falkoschindler** [reply]
@vlogoth This is a bit confusing because in 2022 `ui.table` implemented an AG Grid. This has been renamed in `ui.aggrid` and `ui.table` is not a Quasar table. This is why the suggestions from this discussion don't make too much sense nowadays.

The problem with the "clearSelection" method seems to be that Quasar emits an "update:selected" event, but `ui.table` is listening for the "selection" event. I'm not sure why there are two and if `ui.table` should listen for both.

In your case I'd recommend one of the following solutions:
```py
def clear_selection1():
    test.selected = []

def clear_selection2():
    test.selected.clear()
    test.update()

ui.button('Clear selection 1', on_click=clear_selection1)
ui.button('Clear selection 2', on_click=clear_selection2)
```

**Reply by vlogoth** [reply]
Alright thanks for the lightning fast response. 
On principle, I'd say the table should respond to the quasar prop for sake of uniformity.
The suggestions work for me anyways. 
Love what you've done with nicegui!
